File: src/core/filter.ts
import * as fs from "fs";
import * as path from "path";
import type { Args } from "../cli/args.js";
import { isBinaryFile, parseSize } from "../utils/files.js";
import { LineCounterError } from "./errors.js";
import { DEFAULT_IGNORED_EXTENSIONS } from "./ignored.js";
import {
  detectProjectType,
  getProjectExcludes,
  type ProjectType,
} from "./project-type.js";

function loadDefaultIgnoredExtensions(): string[] {
  return DEFAULT_IGNORED_EXTENSIONS.map((ext) => ext.toLowerCase());
}

export interface FilterPatterns {
  exclude_patterns: RegExp[];
  exclude_extensions: string[];
  include_extensions: string[];
  exclude_dirs: RegExp[];
  include_dirs: RegExp[];
  exclude_names: RegExp[];
  include_names: RegExp[];
  detected_project_type?: ProjectType;
}

export function createFilterPatterns(
  args: Args,
): FilterPatterns | LineCounterError {
  try {
    if (args.max_size) {
      const maxSize = parseSize(args.max_size);
      if (maxSize instanceof LineCounterError) {
        return maxSize;
      }
    }

    if (args.min_size) {
      const minSize = parseSize(args.min_size);
      if (minSize instanceof LineCounterError) {
        return minSize;
      }
    }

    const detectedProjectType = detectProjectType(args.directory);
    const projectExcludes = getProjectExcludes(detectedProjectType);

    const exclude_patterns: RegExp[] = args.exclude_patterns.map((p) => {
      try {
        return new RegExp(p, args.ignore_case ? "i" : undefined);
      } catch (e) {
        throw LineCounterError.invalidRegex(
          p,
          e instanceof Error ? e : undefined,
        );
      }
    });

    const exclude_extensions: string[] = [
      ...args.exclude_extensions.map((e) => e.replace(/^\./, "").toLowerCase()),
      ...projectExcludes.exclude_extensions.map((e) => e.toLowerCase()),
      ...loadDefaultIgnoredExtensions(),
    ];
    const uniqueExcludeExt = Array.from(new Set(exclude_extensions)).sort();

    const include_extensions: string[] = args.include_extensions.map((e) =>
      e.replace(/^\./, "").toLowerCase(),
    );

    const exclude_dirs_patterns: string[] = [
      ...args.exclude_dirs,
      ...projectExcludes.exclude_dirs,
    ];
    const exclude_dirs: RegExp[] = exclude_dirs_patterns.map((p) => {
      return new RegExp(p, args.ignore_case ? "i" : undefined);
    });

    const include_dirs: RegExp[] = args.include_dirs.map((p) => {
      return new RegExp(p, args.ignore_case ? "i" : undefined);
    });

    const exclude_names_patterns: string[] = [
      ...args.exclude_names,
      ...projectExcludes.exclude_names,
    ];
    const exclude_names: RegExp[] = exclude_names_patterns.map((p) => {
      return new RegExp(p, args.ignore_case ? "i" : undefined);
    });

    const include_names: RegExp[] = args.include_names.map((p) => {
      return new RegExp(p, args.ignore_case ? "i" : undefined);
    });

    return {
      exclude_patterns,
      exclude_extensions: uniqueExcludeExt,
      include_extensions,
      exclude_dirs,
      include_dirs,
      exclude_names,
      include_names,
      detected_project_type: detectedProjectType,
    };
  } catch (e) {
    if (e instanceof LineCounterError) {
      return e;
    }
    return LineCounterError.io(
      `Failed to create filter patterns: ${e instanceof Error ? e.message : String(e)}`,
    );
  }
}

export function shouldExcludeFile(
  filePath: string,
  args: Args,
  patterns: FilterPatterns,
): boolean {
  const pathStr = filePath;
  const fileName = path.basename(filePath);

  for (const pattern of patterns.exclude_patterns) {
    if (pattern.test(pathStr)) {
      return true;
    }
  }

  const ext = path.extname(filePath).replace(/^\./, "").toLowerCase();
  if (ext) {
    for (const excludeExt of patterns.exclude_extensions) {
      if (ext === excludeExt.toLowerCase()) {
        return true;
      }
    }

    if (patterns.include_extensions.length > 0) {
      let matches = false;
      for (const includeExt of patterns.include_extensions) {
        if (ext === includeExt.toLowerCase()) {
          matches = true;
          break;
        }
      }
      if (!matches) {
        return true;
      }
    }
  } else if (patterns.include_extensions.length > 0) {
    return true;
  }

  const parentDir = path.dirname(filePath);
  for (const pattern of patterns.exclude_dirs) {
    if (pattern.test(parentDir)) {
      return true;
    }
  }

  if (patterns.include_dirs.length > 0) {
    let matches = false;
    for (const pattern of patterns.include_dirs) {
      if (pattern.test(parentDir)) {
        matches = true;
        break;
      }
    }
    if (!matches) {
      return true;
    }
  }

  for (const pattern of patterns.exclude_names) {
    if (pattern.test(fileName)) {
      return true;
    }
  }

  if (patterns.include_names.length > 0) {
    let matches = false;
    for (const pattern of patterns.include_names) {
      if (pattern.test(fileName)) {
        matches = true;
        break;
      }
    }
    if (!matches) {
      return true;
    }
  }

  if (args.no_hidden && fileName.startsWith(".")) {
    return true;
  }

  try {
    const stats = fs.statSync(filePath);
    const size = stats.size;

    if (args.max_size) {
      const maxSize = parseSize(args.max_size);
      if (!(maxSize instanceof LineCounterError) && size > maxSize) {
        return true;
      }
    }

    if (args.min_size) {
      const minSize = parseSize(args.min_size);
      if (!(minSize instanceof LineCounterError) && size < minSize) {
        return true;
      }
    }

    if (args.no_empty && size === 0) {
      return true;
    }
  } catch {}

  if (args.no_binary && isBinaryFile(filePath)) {
    return true;
  }

  return false;
}

File: src/core/filter.rs
use crate::cli::Args;
use crate::core::{LineCounterError, Result};
use crate::utils;
use regex::Regex;
use std::fs;
use std::path::Path;

pub struct FilterPatterns {
    pub exclude_patterns: Vec<Regex>,
    pub exclude_extensions: Vec<String>,
    pub include_extensions: Vec<String>,
    pub exclude_dirs: Vec<Regex>,
    pub include_dirs: Vec<Regex>,
    pub exclude_names: Vec<Regex>,
    pub include_names: Vec<Regex>,
}

impl FilterPatterns {
    pub fn from_args(args: &Args) -> Result<Self> {
        let exclude_patterns: Vec<Regex> = args
            .exclude_patterns
            .iter()
            .map(|p| {
                let pattern = if args.ignore_case {
                    format!("(?i){}", p)
                } else {
                    p.clone()
                };
                Regex::new(&pattern)
                    .map_err(|e| LineCounterError::InvalidRegex(format!("{}: {}", p, e)))
            })
            .collect::<Result<Vec<_>>>()?;

        let exclude_extensions: Vec<String> = args
            .exclude_extensions
            .iter()
            .map(|e| e.trim_start_matches('.').to_lowercase())
            .collect();

        let include_extensions: Vec<String> = args
            .include_extensions
            .iter()
            .map(|e| e.trim_start_matches('.').to_lowercase())
            .collect();

        let exclude_dirs: Vec<Regex> = args
            .exclude_dirs
            .iter()
            .map(|p| {
                let pattern = if args.ignore_case {
                    format!("(?i){}", p)
                } else {
                    p.clone()
                };
                Regex::new(&pattern).unwrap()
            })
            .collect();

        let include_dirs: Vec<Regex> = args
            .include_dirs
            .iter()
            .map(|p| {
                let pattern = if args.ignore_case {
                    format!("(?i){}", p)
                } else {
                    p.clone()
                };
                Regex::new(&pattern).unwrap()
            })
            .collect();

        let exclude_names: Vec<Regex> = args
            .exclude_names
            .iter()
            .map(|p| {
                let pattern = if args.ignore_case {
                    format!("(?i){}", p)
                } else {
                    p.clone()
                };
                Regex::new(&pattern).unwrap()
            })
            .collect();

        let include_names: Vec<Regex> = args
            .include_names
            .iter()
            .map(|p| {
                let pattern = if args.ignore_case {
                    format!("(?i){}", p)
                } else {
                    p.clone()
                };
                Regex::new(&pattern).unwrap()
            })
            .collect();

        Ok(Self {
            exclude_patterns,
            exclude_extensions,
            include_extensions,
            exclude_dirs,
            include_dirs,
            exclude_names,
            include_names,
        })
    }
}

pub fn should_exclude_file(path: &Path, args: &Args, patterns: &FilterPatterns) -> bool {
    let path_str = path.to_string_lossy();
    let file_name = path.file_name().and_then(|n| n.to_str()).unwrap_or("");

    for pattern in &patterns.exclude_patterns {
        if pattern.is_match(&path_str) {
            return true;
        }
    }

    if let Some(ext) = path.extension().and_then(|e| e.to_str()) {
        let ext_lower = ext.to_lowercase();
        for exclude_ext in &patterns.exclude_extensions {
            if ext_lower == exclude_ext.to_lowercase() {
                return true;
            }
        }

        if !patterns.include_extensions.is_empty() {
            let mut matches = false;
            for include_ext in &patterns.include_extensions {
                if ext_lower == include_ext.to_lowercase() {
                    matches = true;
                    break;
                }
            }
            if !matches {
                return true;
            }
        }
    } else if !patterns.include_extensions.is_empty() {
        return true;
    }

    for pattern in &patterns.exclude_dirs {
        if let Some(parent) = path.parent() {
            if pattern.is_match(&parent.to_string_lossy()) {
                return true;
            }
        }
    }

    if !patterns.include_dirs.is_empty() {
        let mut matches = false;
        if let Some(parent) = path.parent() {
            for pattern in &patterns.include_dirs {
                if pattern.is_match(&parent.to_string_lossy()) {
                    matches = true;
                    break;
                }
            }
        }
        if !matches {
            return true;
        }
    }

    for pattern in &patterns.exclude_names {
        if pattern.is_match(file_name) {
            return true;
        }
    }

    if !patterns.include_names.is_empty() {
        let mut matches = false;
        for pattern in &patterns.include_names {
            if pattern.is_match(file_name) {
                matches = true;
                break;
            }
        }
        if !matches {
            return true;
        }
    }

    if args.no_hidden && file_name.starts_with('.') {
        return true;
    }

    if let Ok(metadata) = fs::metadata(path) {
        let size = metadata.len();

        if let Some(ref max_size_str) = args.max_size {
            if let Ok(max_size) = utils::parse_size(max_size_str) {
                if size > max_size {
                    return true;
                }
            }
        }

        if let Some(ref min_size_str) = args.min_size {
            if let Ok(min_size) = utils::parse_size(min_size_str) {
                if size < min_size {
                    return true;
                }
            }
        }

        if args.no_empty && size == 0 {
            return true;
        }
    }

    if args.no_binary && utils::is_binary_file(path) {
        return true;
    }

    false
}
